### PR TITLE
#159965320 allow users to bookmark articles 

### PR DESCRIPTION
**Pull request by meshack-mbuvi**
#### What does this pull request do?
- Enable users to bookmark an article
- Enable users to retrieve and delete their bookmarks

#### Description of tasks to be completed
- Ensure to have tests for bookmarking, retrieving and deleting an article bookmark
- Add views, serializers, and URLs to bookmark, retrieve and delete an article bookmark

#### How should this be manually tested?
- Clone [this](https://github.com/andela/ah-magnificent6.git) repository
- Copy everything on *.env-sample* to  *.env* [and set the environment variables]
- Create and activate a [virtual environment](https://docs.python.org/3/library/venv.html) inside the project base directory.
- Run the command `git checkout ft-bookmark-articles-159965320`
- Install the project dependencies by running `pip install -r requirements.txt`
- Run the server by issuing the command `python manage.py runserver`
- Using ([Postman](https://www.getpostman.com/) or your favorite REST API client, 
- Create an account by accessing `api/users/signup/` by sending the payload in the format`
`{
  "firstname": "admin",
  "username": "admin",
  "email": "admin@gmail.com",
  "password": "password",
} 
`
and 
login on `/api/users/login/` using the same format for the payload above
- test the above endpoints by passing data in the format 
`{
    'title':'title of the article',
    'body':'body of the article',
    'description':'Description of the article'
}`
- Now that you have created an article, visit `http://127.0.0.1:8000/api/articles/slug/bookmarks` to bookmark the article created above
- To retrieve or delete your bookmarks, visit`http://127.0.0.1:8000/api/articles/bookmarks/{bookmark_id}` using `get` method to retrieve a bookmark or `delete` method to delete your bookmark.
**Note** Visiting `http://127.0.0.1:8000/api/articles/bookmarks/` using `get` retrieves all your bookmarks while using `delete` deletes all your bookmarks

#### What are the relevant pivotal tracker stories?
[#159965320 Users can bookmark an article](https://www.pivotaltracker.com/story/show/159965320)

#### Screenshots
image
image
image

### CODE REVIEW

**Comment by micahoriaso** [comment]
screen shot 2018-09-24 at 12 13 07
@meshack-mbuvi your response message  for a GET request for all bookmarks is not properly formatted. Please fix this

**Comment by meshack-mbuvi** [comment]
@micahoriaso I have already done so

**Comment by micahoriaso** [comment]
Looks good to me

**Comment by mikenthiwa** [comment]
@meshack-mbuvi Check how we can represent the user as username other than user id

**Comment by Bileonaire** [comment]
LOOKS GOOD TO ME. Kindly resolve the return when one creates the bookmark not to return the user id. You can do that in the serializer.

**Comment by meshack-mbuvi** [comment]
@Bileonaire confirm that have attended to your request.
image

**Comment by Bileonaire** [comment]
@meshack-mbuvi It is okay now

**Comment by mikenthiwa** [comment]
@meshack-mbuvi looks good, Improve on test coverage

File: authors/apps/articles/views.py [hunk 1]
```diff
@@ -639,3 +637,89 @@ def delete(self, request, slug, pk):
...
+                }, status.HTTP_404_NOT_FOUND
+            )
+
+
+class ArticleBookmarkDetailAPIView(generics.RetrieveDestroyAPIView):
+    """
+    get:
+    Retrieve all bookmarks for a logged in user
+    """
+    permission_classes = (IsAuthenticatedOrReadOnly, )
```
Comment (parseen254) [inline authors/apps/articles/views.py]: This should be for Authenticated Users only @meshack-mbuvi

File: authors/apps/articles/views.py [hunk 2]
```diff
@@ -639,3 +637,89 @@ def delete(self, request, slug, pk):
             return Response(data={
                 'message': 'That article report does not exist.'
             }, status=status.HTTP_404_NOT_FOUND)
+class ArticleBookmarkAPIView(generics.CreateAPIView):
+    """
+    post:
+    Bookmark an article
+    """
+    renderer_classes = (BookmarkJSONRenderer, )
+    permission_classes = (IsAuthenticatedOrReadOnly, )
```
Comment (parseen254) [inline authors/apps/articles/views.py]: Let's keep this for Authenticated users only.

**Review by parseen254 — CHANGES_REQUESTED** [review]

**Review by parseen254 — APPROVED** [review]

**Pull request merged**